pkg/apis/metrics/types.go

```diff
@@ -16,14 +16,205 @@ limitations under the License.
+type NetworkMetrics struct {
+	// Cumulative count of bytes received.
+	RxBytes *resource.Quantity `json:"rxBytes,omitempty"`
+	// Cumulative count of receive errors encountered.
+	RxErrors *int64 `json:"rxErrors,omitempty"`
+	// Cumulative count of bytes transmitted.
+	TxBytes *resource.Quantity `json:"txBytes,omitempty"`
+	// Cumulative count of transmit errors encountered.
+	TxErrors *int64 `json:"txErrors,omitempty"`
+}
+
+// CPUMetrics contains data about CPU usage.
+type CPUMetrics struct {
+	// Total CPU usage (sum of all cores) averaged over the sample window.
+	// The "core" unit can be interpreted as CPU core-seconds per second.
+	TotalCores *resource.Quantity `json:"totalCores,omitempty"`
+	// CPU load that the container is experiencing, represented as a smoothed average of number of
+	// runnable threads.  Load is averaged over the sampling window.
+	LoadAverage *resource.Quantity `json:"loadAverage,omitempty"`
+}
+
+// MemoryMetrics contains data about memory usage.
+type MemoryMetrics struct {
+	// Total memory in use. This includes all memory regardless of when it was accessed.
+	TotalBytes *resource.Quantity `json:"totalBytes,omitempty"`
+	// The amount of working set memory. This includes recently accessed memory,
+	// dirty memory, and kernel memory. UsageBytes is <= TotalBytes.
+	UsageBytes *resource.Quantity `json:"usageBytes,omitempty"`
+	// Cumulative number of minor page faults.
+	PageFaults *int64 `json:"pageFaults,omitempty"`
+	// Cumulative number of major page faults.
+	MajorPageFaults *int64 `json:"majorPageFaults,omitempty"`
+}
+
+// CustomMetricType specifies the type of metric being exported.
+type CustomMetricType string
+
+const (
+	// Instantaneous value. May increase or decrease.
+	CustomMetricGauge CustomMetricType = "Gauge"
+	// A counter-like value that increases monotonically.
+	CustomMetricCumulative CustomMetricType = "Cumulative"
+	// Rate over a time period.
+	CustomMetricDelta CustomMetricType = "Delta"
+)
+
+// CustomMetric provides custom application metric data.
+// See https://github.com/google/cadvisor/blob/master/docs/application_metrics.md
+type CustomMetric struct {
+	// The name of the metric.
+	Name string `json:"name"`
+	// Type of the metric.
+	Type CustomMetricType `json:"type"`
+	// Display Unit for the sample values.
+	Unit string `json:"unit"`
+	// Historical custom metric samples.
+	Samples []CustomMetricSample `json:"samples,omitempty" patchStrategy:"merge" patchMergeKey:"sampleTime"`
+}
+
+// CustomMetric contains a single sample point for a custom metric.
+type CustomMetricSample struct {
+	Sample `json:",inline"`
+	// Label associated with a metric
+	Label *string `json:"label,omitempty"`
+	// The value of the metric at this point.
+	// The unit is defined by the CustomMetric this sample belongs to.
+	Value resource.Quantity `json:"value"`
+}
+
+// RawMetricsOptions are the query options for raw metrics endpoints.
+type RawMetricsOptions struct {
```

Comment:
If you are planning to provide a Watch-like api then there should be a way to express when to send the next set of metrics. I would like to aviod the situation when all Kubelets push their new data to Heapster at the same time.